java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.lancedb.lance.spark.write;
+
+import com.lancedb.lance.FragmentMetadata;
+import com.lancedb.lance.spark.LanceConfig;
+import com.lancedb.lance.spark.internal.LanceDatasetAdapter;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.write.DataWriter;
+import org.apache.spark.sql.connector.write.DataWriterFactory;
+import org.apache.spark.sql.connector.write.WriterCommitMessage;
+import org.apache.spark.sql.types.StructType;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+
+public class LanceDataWriter implements DataWriter<InternalRow> {
+  private InternalRowWriterArrowReader arrowReader;
+  private FutureTask<FragmentMetadata> fragmentCreationTask;
+  private Thread fragmentCreationThread;
+
+  private LanceDataWriter(InternalRowWriterArrowReader arrowReader,
+      FutureTask<FragmentMetadata> fragmentCreationTask, Thread fragmentCreationThread) {
+    // TODO support write to multiple fragments
+    this.arrowReader = arrowReader;
+    this.fragmentCreationThread = fragmentCreationThread;
+    this.fragmentCreationTask = fragmentCreationTask;
+  }
+
+  @Override
+  public void write(InternalRow record) throws IOException {
+    arrowReader.write(record);
+  }
+
+  @Override
+  public WriterCommitMessage commit() throws IOException {
```

Comment:
This is commit at at the end of the full write? If so, we will accumulate too much data in `arrowReader`.

Reply:
The current sequence is
- create arrow reader
- start the reader thread and execute Fragment.create(reader)
- write to the arrow reader
- reader thread (fragment.create) consumes data in batches
- commit which notify there is no more new data and wait for reader thread (fragment.create) finishes